tools/azure-sdk-tools/devtools_testutils/azure_testcase.py

```diff
@@ -182,34 +187,52 @@ def setUp(self):
     def tearDown(self):
         return super(AzureTestCase, self).tearDown()
 
-    def create_basic_client(self, client_class, **kwargs):
 
-        tenant_id = os.environ.get("AZURE_TENANT_ID", None)
-        client_id = os.environ.get("AZURE_CLIENT_ID", None)
-        secret = os.environ.get("AZURE_CLIENT_SECRET", None)
 
-        if tenant_id and client_id and secret and self.is_live:
+    def _create_credential(self, client_secret_credential):
+        return client_secret_credential(
+            tenant_id=self.tenant_id,
+            client_id=self.client_id,
+            client_secret=self.secret
+        )
+
+    def _get_real_credential(self, is_async):
+        from azure.identity import ClientSecretCredential
+        if is_async:
+            from azure.identity.aio import ClientSecretCredential
+        return self._create_credential(ClientSecretCredential)
+
+    def _get_fake_credential(self, is_async):
+        if is_async:
+            return AsyncFakeCredential()
+        return self.settings.get_azure_core_credentials()
+
+    def get_credential(self, client_class, **kwargs):
+
+        self.tenant_id = os.environ.get("AZURE_TENANT_ID", None)
+        self.client_id = os.environ.get("AZURE_CLIENT_ID", None)
+        self.secret = os.environ.get("AZURE_CLIENT_SECRET", None)
+        is_async = kwargs.pop("is_async", False)
+
+        if self.tenant_id and self.client_id and self.secret and self.is_live:
             if _is_autorest_v3(client_class):
                 # Create azure-identity class
-                from azure.identity import ClientSecretCredential
-                credentials = ClientSecretCredential(
-                    tenant_id=tenant_id,
-                    client_id=client_id,
-                    client_secret=secret
-                )
+                return self._get_real_credential(is_async)
             else:
                 # Create msrestazure class
                 from msrestazure.azure_active_directory import ServicePrincipalCredentials
-                credentials = ServicePrincipalCredentials(
+                return ServicePrincipalCredentials(
                     tenant=tenant_id,
                     client_id=client_id,
                     secret=secret
                 )
```

Comment:
as now you're approaching the `_get_real_credential` method for returning real credential, I'm wondering we could do further unification here for this `ServicePrincipalCredentials` as well.

maybe we could just tweak the `_get_real_credential` signature a bit, making it
```
def _get_real_credential(self, is_async=false, crendetial_type=None):
# if crendetial_type is provided, just use the given type
if crendetial_type:
   return crendetial_type(client_id=...)
# otherwise we fall back to some default paths
if is_async:
    from ..
```

Reply:
@yunhaoling that would have to be a string, since we can't import aio code in 2.7 (right now it works because the aio code is smartly wrapped in the right if).

So I'm not against, but I'm not sure we need in the first place (are we trying to solve an actual problem, or just do something cool for fun? :)) and this string challenge make the API less than ideal.

I'm happy to be pushed back if you strongly believe in it

Reply:
@lmazuel emmm, I see the reverse, it's because we're in the py2.7 so that we won't be importing and passing aio credential_type into the method -- it's the caller's duty to check which env/is async/etc. and make the right import.
the callee `_get_credential` simply takes what has been passed.

as you said, it's mostly "something cool for fun" -- the place where I come from is for the potential long term maintainability/scalability/readability -- that's what a unified approach could potentially provide.

but I might be bikeshedding and overengineering, let's waste no more time on this and get it work in our repo first😊